sdk/src/main/scala/com/horizen/account/state/AccountState.scala

```diff
@@ -352,21 +352,29 @@ class AccountState(val params: NetworkParams,
   override def getBlockGasLimit: BigInteger = using(getView)(_.getBlockGasLimit)
 
   override def validate(tx: SidechainTypes#SCAT): Try[Unit] = Try {
-    tx.semanticValidity()
+    Try(tx.semanticValidity()) recoverWith {
+      case e: Throwable =>
+        log.debug("Transaction is not semantically valid" , e)
+        Failure(e)
+    }  get
 
     if (tx.isInstanceOf[EthereumTransaction]) {
-
-      val ethTx = tx.asInstanceOf[EthereumTransaction]
-      val blockGasPool = new GasPool(getBlockGasLimit)
-
       using(getView) { stateView =>
-        stateView.applyTransaction(tx, 0, blockGasPool) match {
```

Comment:
It seems dangerous that we removed the full validation of Tx.
I'm pretty sure that we are not safe now.
Maybe we need to add parameter to the applyTransaction like "skip tx nonce too high"?
to be discussed.

Reply:
In theory, all the checks needed for deciding if a tx is valid or not must be performed when a tx is added to the mempool and when a block is applied to the blockchain. In addition, some checks related to the state (as balance and nonce), must be performed again during forging and when the mempool is reorganized after the blockchain has been modified.
All these checks can be performed in one place, i.e. applyTransaction, but this would mean executing the transaction when is added to the mempool, that seems too much.
Now the checks are performed in different methods (eg semanticValidity, validate, precheck) and the result should be the same as if applyTransaction was used.
The only thing missed is the case a bug is inside applyTransaction that causes a tx to be considered invalid. In this case, the tx is not included in a block or the block is discarded, but the tx will be however included in the mempool.